src/main/java/org/jitsi/videobridge/LipSyncHack.java

```diff
@@ -0,0 +1,423 @@
+     * and injection.
+     */
+    private final ScheduledExecutorService scheduler =
+        Executors.newScheduledThreadPool(1);
+
+    /**
+     * The remote audio SSRCs that have been accepted by the translator and
+     * forwarded to the endpoint associated to this instance.
+     */
+    private final List<Long> acceptedAudioSSRCs = new ArrayList<>();
+
+    /**
+     * The remote video SSRCs that have been accepted by the translator and
+     * forwarded to the endpoint associated to this instance.
+     */
+    private final List<Long> acceptedVideoSSRCs = new ArrayList<>();
+
+    /**
+     * A map that holds all the inject states.
+     */
+    private final Map<Long, InjectState> states = new HashMap<>();
+
+    /**
+     * Ctor.
+     *
+     * @param endpoint
+     */
+    public LipSyncHack(Endpoint endpoint)
+    {
+        this.endpoint = endpoint;
+    }
+
+    /**
+     * Notifies this instance that an audio packet is about to be written.
+     *
+     * @param data
+     * @param buffer
+     * @param offset
+     * @param length
+     * @param source
+     */
+    public void onRTPTranslatorWillWriteAudio(
+        boolean data, byte[] buffer, int offset,
+        int length, Channel source)
+    {
+        if (!data)
+        {
+            return;
+        }
+
+        // Decide whether to trigger the hack or not.
+        Long acceptedAudioSSRC
+            = RawPacket.getSSRCAsLong(buffer, offset, length);
+
+        // In order to minimize the synchronization overhead, we process
+        // only the first data packet of a given RTP stream.
+        //
+        // XXX No synchronization is required to r/w the acceptedAudioSSRCs
+        // because in the current architecture this method is called by a single
+        // thread at the time.
+        if (acceptedAudioSSRCs.contains(acceptedAudioSSRC))
+        {
+            // We've already triggered the hack for this audio stream and its
+            // associated video streams.
+            return;
+        }
+
+        acceptedAudioSSRCs.add(acceptedAudioSSRC);
+
+        // New audio stream. Trigger the hack for the associated video stream.
+        List<RtpChannel> targetVideoChannels
+            = endpoint.getChannels(MediaType.VIDEO);
+        if (targetVideoChannels == null || targetVideoChannels.size() == 0)
+        {
+            return;
+        }
+
+        VideoChannel targetVC = (VideoChannel) targetVideoChannels.get(0);
+        if (targetVC == null)
+        {
+            return;
+        }
+
+        List<RtpChannel> sourceVideoChannels
+            = source.getEndpoint().getChannels(MediaType.VIDEO);
+        if (sourceVideoChannels == null || sourceVideoChannels.size() == 0)
+        {
+            return;
+        }
+
+        VideoChannel sourceVC = (VideoChannel) sourceVideoChannels.get(0);
+        if (sourceVC == null)
+        {
+            return;
+        }
+
+        // FIXME this will include rtx ssrcs and whatnot.
```

Comment:
We create a state object for a given video SSRC only once. So, however many audio SSRCs we receive from the a given audio channel, the SSRCs from the associated video channel are added only once.